Ical.Net/CollectionExtensions.cs

```diff
@@ -0,0 +1,58 @@
+
+namespace Ical.Net;
+
+public static class CollectionExtensions
+{
+    /// <summary>
+    /// Returns the elements of the sequence of occurrences to only include those that start before the specified period end.
+    /// Important: The input sequence <b>must be ordered</b> according to the type's default equality comparer.
+    /// <para/>
+    /// A perfect fit is the sequence returned by <see cref="RecurringComponent.GetOccurrences(CalDateTime, EvaluationOptions)"/>,
+    /// like the frequently used <see cref="CalendarEvent.GetOccurrences(CalDateTime, EvaluationOptions)"/>
+    /// or <see cref="Calendar.GetOccurrences"/>.
+    /// <para/>
+    /// This is a convenience method for <see cref="Enumerable.TakeWhile{TSource}(IEnumerable{TSource}, Func{TSource, bool})"/>.
+    /// But even with this method, the input sequence must be ordered according to the type's default equality comparer.
+    /// </summary>
+    /// <param name="sequence">The ordered sequence to be filtered.</param>
+    /// <param name="periodEnd">The exclusive end of the period to be used as a filter.</param>
+    /// <returns>
+    /// The elements of the sequence of occurrences to only include those that start before the specified period end.
+    /// </returns>
+    public static IEnumerable<Occurrence> TakeBefore(this IEnumerable<Occurrence> sequence, CalDateTime? periodEnd)
+        => (periodEnd == null) ? sequence : sequence.TakeWhile(p => p.Period.StartTime < periodEnd);
+
+    /// <summary>
+    /// Returns the elements of the sequence of periods to only include those that start before the specified period end.
+    /// Important: The input sequence <b>must be ordered</b> according to the type's default equality comparer.
+    /// <para/>
+    /// A perfect fit is the sequence returned by the <see cref="Evaluator.Evaluate"/> implementations
+    /// like <see cref="RecurrencePatternEvaluator.Evaluate"/>.
+    /// <para/>
+    /// This is a convenience method for <see cref="Enumerable.TakeWhile{TSource}(IEnumerable{TSource}, Func{TSource, bool})"/>
+    /// <param name="sequence">The ordered sequence to be filtered.</param>
+    /// <param name="periodEnd">The exclusive end of the period to be used as a filter.</param>
+    /// <returns>
+    /// The elements of the sequence of periods to only include those that start before the specified period end.
+    /// But even with this method, the input sequence must be ordered according to the type's default equality comparer.
+    /// </returns>
+    /// </summary>
+    /// <param name="sequence">The ordered sequence to be filtered.</param>
+    /// <param name="periodEnd">The exclusive end of the period to be used as a filter.</param>
+    /// <returns>
+    /// The elements of the sequence of periods to only include those that start before the specified period end.
+    /// </returns>
+    public static IEnumerable<Period> TakeBefore(this IEnumerable<Period> sequence, CalDateTime? periodEnd)
```

Comment:
Regarding the naming: `TakeBefore` could be interpreted as filter (aka 'take all items from the sequence where the value is _before_ x). In this interpretation it wouldn't reflect the fact that 'Before' refers to the items in the sequence, which means _take those that are returned before hitting or exceeding the threshold). An alternative could be `TakeWhileBefore`, which more clearly suggests 'take all items from the sequence while the their date value is before x', but the name is a bit clumsy. Not sure.

Reply:
Yes, really makes sense to think about the best term. Coming back to one of the initial ideas using like **`.GetOccurrences(...).WhileBefore(...)`** sounded quite natural, and it was short... Better?

Reply:
Sounds better to me. It's not quite in line with linq naming though. Not sure it has to be in this case. On the other hand, if it is named like TakeWhileXxx it could be easier to find, because it pops up in auto completion when people are looking for linq operators. Anyone, any of the discussed names is good.

Reply:
"easier to find" is convincing for `.TakeWhileBefore(...)`, tnx